Adding more accurate information about client side scripts (#522)
withastro/docs: src/pages/en/core-concepts/astro-components.md

To send JavaScript to the browser without [using a framework component](/en/core-concepts/framework-components) (React, Svelte, Vue, Preact, SolidJS, AlpineJS, Lit) or an [Astro integration](https://astro.build/integrations/) (e.g. astro-XElement), you can use a `<script>` tag in your Astro component template and send JavaScript to the browser that executes in the global scope.

> By default, `<script>` tags are hoisted and `type="module"`. They will be placed in the `<head>` of the document and will be bundled with your imports. This is useful for sending client-side scripts to the browser that are not tied to a specific component.

Reviewer: Do we need to include some help with understanding ES Modules? Maybe a link like: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules

There’s also this incredible deep dive by Lin Clark, but probably not helpful in this context: https://hacks.mozilla.org/2018/03/es-modules-a-cartoon-deep-dive/

> By default, `<script>` tags are hoisted and `type="module"`. They will be placed in the `<head>` of the document and will be bundled with your imports. This is useful for sending client-side scripts to the browser that are not tied to a specific component.

Author: Maybe add the link to Mozilla docs on the `type="module"` statement

> By default, `<script>` tags are hoisted and `type="module"`. They will be placed in the `<head>` of the document and will be bundled with your imports. This is useful for sending client-side scripts to the browser that are not tied to a specific component.

Reviewer: Sounds good!

> By default, `<script>` tags are hoisted and `type="module"`. They will be placed in the `<head>` of the document and will be bundled with your imports. This is useful for sending client-side scripts to the browser that are not tied to a specific component.

Reviewer 2: Love this. Happy. You two got this one.

> > ⚠️ TypeScript support is not currently available for client-side scripts, but we [are discussing](https://github.com/withastro/rfcs/discussions/193) adding this feature.

Reviewer: I’m not sure we usually include links to RFCs in the docs? Maybe we could say this instead:

```suggestion
> ⚠️ You can’t currently write TypeScript in client-side scripts, but you _can_ import a Typescript file if you prefer writing with that syntax.
```

Author: If I remember correctly Fred used the RFC link on the docs (I can't remember which page) to endorse users to check out by themselves the current status

Reviewer 2: You're both right! 💜  @Chrissdroid @delucis 

Fred has used RFC's as placeholders in the past, but they really are placeholders and are updated once there's something better/official to link them to. So, it's not unheard of, but perhaps best left to things that:

-  *need* but don't yet have any documentation (I think that was the case in the past)
- are realistically coming soon? While well received, this RFC is only a few days old. So even if agreed-upon, I'm not sure how long it would take to make it happen?
- are key/core features that people expect or need, and might be put off by it not existing yet

I'm OK either way, link or no link, as long as we think it's reasonably justified by a criteria like above. Otherwise, the docs could be full of links to things in various stages of development, that may or may not happen, and then require US to manually keep track of all these things and update pages separately from these RFCs being closed.

Reviewer: I think we only link to _accepted_ RFCs for more information (like on deprecating the Markdown component). I still feel providing the workaround here (that you can write a `.ts` file and import that) is probably more helpful than the RFC.

Author: Agree with @delucis

> > ⚠️ TypeScript support is not currently available for client-side scripts. Please a look at the [RFC](https://github.com/withastro/rfcs/discussions/193) for more information on the status of this feature.

Reviewer 2: ```suggestion
> ⚠️ TypeScript support is not currently available for client-side scripts, but we [are discussing](https://github.com/withastro/rfcs/discussions/193) adding this feature.
```

```astro
<script>
  // Processed! Bundled! ESM imports work, even to npm packages.
</script>
```

> In some cases, you may avoid bundling the script and using it as it is via the `is:inline` attribute.

Reviewer 2: ```suggestion
To avoid bundling the script, you can use the `is:inline` attribute.
```

```astro
<script is:inline>
  // Will be rendered into the HTML exactly as written!
  // ESM imports will not be resolved relative to the file.
</script>
```

Multiple `<script>` tags can be used on the same Astro component with methods above.

> > **Note :** Adding `type="module"` or any other attribute to the `<script>` tag will prevent it from being bundled by Astro.

Reviewer 2: ```suggestion
> **Note :** Adding`type="module"` or any other attribute to  a `<script>` tag removes the default hoist behaviour and  prevents it from being bundled by Astro.
```
I'm not sure if this is the best way to say "removes the hoist", but I think it's worth saying that:
- it removes the hoise
-  it's now just like you've added is:inline
Can you think of the best way to get those words in there?  :)

Author: ```suggestion
> **Note :** Adding `type="module"` or any other attribute to a `<script>` tag will treats it it's `is:inline` directive and remove it's default `hoist` behavior, disabling bundling for this specific tag.
```

More like this ?

Reviewer 2: Yes, more like that.  :)  I'm going to ask @delucis to help with the exact wording, because my brain is a little mushy, but yes, I think that, and the rest of this, is exactly what we needed for the small scope of improving this page right now! 🥳

Reviewer: ```suggestion
> **Note:** Adding `type="module"` or any other attribute to a `<script>` tag will cause it to be treated as if it had an `is:inline` directive and remove its default behavior, disabling bundling for the tag.
```

Reviewer 2: > **Note:** Adding `type="module"` or any other attribute to a `<script>` tag will **disable** Astro's default bundling behavior, treating it as if it had an `is:inline` directive.

Does this work?

**disable? override? revert? ignore?**

Reviewer: 💯 **disable** works perfectly \*chef’s kiss\*

Author: From the list, i prefer disable with the context given

By default, `<script>` tags are processed by Astro.

- Any imports will be bundled, allowing you to import local files or Node modules.
- The processed script will be injected into your page’s `<head>` with [`type="module"`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules).
- If your component is used several times on a page, the script tag will only be included once.

> ⚠️ You can’t currently write TypeScript in client-side scripts, but you _can_ import a Typescript file if you prefer writing with that syntax.

```astro
<script>
  // Processed! Bundled! ESM imports work, even to npm packages.
</script>
```

To avoid bundling the script, you can use the `is:inline` attribute.

```astro
<script is:inline>
  // Will be rendered into the HTML exactly as written!
  // ESM imports will not be resolved relative to the file.
</script>
```

> Multiple `<script>` tags can be used on the same Astro component with methods above.

Reviewer: Small tweak, but then this LGTM!

```suggestion
Multiple `<script>` tags can be used in the same `.astro` file using any combination of the methods above.
```

Author: It's sooo much better said than mine 😆

Reviewer: I said it was small! 🤣

Reviewer 2: Haha... I was just about to write pretty much exactly this, then it disappeared (resolved) on me!  👍
